### PR TITLE
[SPARK-26191][SQL] Control truncation of Spark plans via maxFields parameter

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to add `maxFields` parameter to all functions involved in creation of textual representation of spark plans such as `simpleString` and `verboseString`. New parameter restricts number of fields converted to truncated strings. Any elements beyond the limit will be dropped and replaced by a `"... N more fields"` placeholder. The threshold is bumped up to `Int.MaxValue` for `toFile()`.  

## How was this patch tested?

Added a test to `QueryExecutionSuite` which checks `maxFields` impacts on number of truncated fields in `LocalRelation`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
ping @hvanhovell

**Comment by MaxGekk** [comment]
@gatorsmile @cloud-fan Could you look at the changes - extracted from another PR: https://github.com/apache/spark/pull/22429

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@HyukjinKwon @dongjoon-hyun @srowen @zsxwing Do you have any objections of this PR?

**Comment by srowen** [comment]
Rather than change every single call to this method, if this should generally be the value of the argument, then why not make it the default value or something?

**Comment by MaxGekk** [comment]
> Rather than change every single call to this method, if this should generally be the value of the argument, then why not make it the default value or something?

New parameter aims to solve the problem when there are multiple callers, and each of them needs different maximum fields. So, a feasible approach is to propagate `maxFields` from callers to `truncatedString`. Changing global SQL config does not solve the problem.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1777,7 +1777,7 @@ class Analyzer(
 
       case p if p.expressions.exists(hasGenerator) =>
         throw new AnalysisException("Generators are not supported outside the SELECT clause, but " +
-          "got: " + p.simpleString)
+          "got: " + p.simpleString((SQLConf.get.maxToStringFields)))
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Nit: are there extra parens here?

**Review by srowen — COMMENTED** [review]
Ah OK so not every call would just pass the value `SQLConf.get.maxToStringFields`? It looked like it from the code here but I didn't examine the whole diff. If it were really always `SQLConf.get.maxToStringFields` and configured that way this could be simpler, but I suppose it isn't.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: fixed here and other places

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan and @gatorsmile .

**Comment by HyukjinKwon** [comment]
Looks okay but let me leave it to @cloud-fan, @gatorsmile and @hvanhovell

**Comment by MaxGekk** [comment]
@hvanhovell We all are waiting for your decision ;-). Please, review the PR.

**Comment by MaxGekk** [comment]
I am going to close the PR since nobody interested in it :-(

**Comment by hvanhovell** [comment]
@MaxGekk can you resolve the merge conflicts?

**Comment by hvanhovell** [comment]
LGTM

**Comment by hvanhovell** [comment]
merging to master.

**Pull request closed without merging**